### PR TITLE
displaycal: init at 3.9.10

### LINKED ISSUE #177623: DisplayCAL Python 3
**Project description**

.`aliases.nix`
```nix
throw "displaycal has been removed from nixpkgs, as it hasn't migrated to python3"; # Added 2022-01-12
```

A few months back the project was removed because Python 2 was removed, and the Pillow lib was insecure too. This project is a revival of the project but under Python 3. In fact they have stated: "Everything should be working now".

DisplayCAL is a tool for many color-related tasks; from the docs:

* Calibration + Characterization (Profiling).
* Installing the created ICC profile both locally and system-wide (requires root permissions).
* Profile Info window is now fully working (on some systems we still have an issue related to default values https://github.com/eoyilmaz/displaycal-py3/issues/67).
* Measurement report creation.
* Creating, displaying and uploading Colorimeter Corrections.
* Measuring and reporting display uniformity.
* Creating charts with Test Chart Editor and creating diagnostic 3d data.
* Creating 3D LUTs.
* Creating synthetic ICC profiles.

**Metadata**

* homepage URL: https://github.com/eoyilmaz/displaycal-py3
* source URL: https://github.com/eoyilmaz/displaycal-py3
* license: GPL 3.0
* platforms: unsure, but at least Linux

### PR DESCRIPTION
**Pull request by Kranzes**
###### Description of changes

Closes https://github.com/NixOS/nixpkgs/issues/177623

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/graphics/displaycal/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/graphics/displaycal/default.nix:46]: Can we use xvfb-run?

Comment (Kranzes) [reply pkgs/applications/graphics/displaycal/default.nix]: I re-added to the comment on why this isn't possible.

**Comment by toastal** [comment]
```diff
diff --git a/pkgs/applications/graphics/displaycal/default.nix b/pkgs/applications/graphics/displaycal/default.nix
index 71956181242..646751b0577 100644
--- a/pkgs/applications/graphics/displaycal/default.nix
+++ b/pkgs/applications/graphics/displaycal/default.nix
@@ -45,9 +45,16 @@ buildPythonApplication rec {
 
   doCheck = false; # Tests try to access an X11 session.
 
+  # no idea why it looks there - symlink .json lang (everything)
   postInstall = ''
+    for x in $out/share/DisplayCAL/*; do
+      ln -s $x $out/lib/python3.10/site-packages/DisplayCAL
+    done
+
     for binary in "$out/bin/"*; do
-      wrapProgram "$binary" --prefix PATH : ${lib.makeBinPath [ argyllcms ]}
+      wrapProgram "$binary" \
+        --prefix PYTHONPATH : "$PYTHONPATH" \
+        --prefix PATH : ${lib.makeBinPath [ argyllcms ]}
     done
   '';
```

(thanks GitHub for not allowing `*.patch` files from being uploaded 🙄)

After applying this patch, similar to the Python 2 version of DisplayCAL, I was able to get this to *mostly* work--at least the critical parts of measuring, calibrating, generating ICC profiles. I did have some residual errors when trying to load anything related to the file systems (such as deleting a profile from the UI or attempting to move the path it saves profiles to).

```console
(.displaycal-wrapped:7566): GLib-GIO-ERROR **: 11:19:22.276: No GSettings schemas are installed on the system
fish: Job 1, 'result/bin/displaycal' terminated by signal SIGTRAP (Trace or breakpoint trap)
```

A web search for the error brought up `nixpkgs` as the first match so possibly a similar issue. https://github.com/NixOS/nixpkgs/issues/72282

**Comment by TheSirC** [comment]
On my side this draft fails with : 
```
    > Executing pipInstallPhase
       > /build/DisplayCAL-3.9.7/dist /build/DisplayCAL-3.9.7
       > Processing ./DisplayCAL-3.9.7-cp310-cp310-linux_x86_64.whl
       > Requirement already satisfied: zeroconf in /nix/store/vz2q3nqgy62jyq0y9arsb2y93pflmwbr-python3.10-zeroconf-0.38.7/lib/python3.10/site-packages (from DisplayCAL==3.9.7) (0.38.7)
       > Requirement already satisfied: numpy in /nix/store/pvl8s3qnppciv9a8z5jh50w32hgbsx0c-python3.10-numpy-1.21.6/lib/python3.10/site-packages (from DisplayCAL==3.9.7) (1.21.6)
       > Requirement already satisfied: Send2Trash in /nix/store/y9dyvhvx6c54bci1m2882sck8a7ngijf-python3.10-Send2Trash-1.8.1b0/lib/python3.10/site-packages (from DisplayCAL==3.9.7) (1.8.1b0)
       > ERROR: Could not find a version that satisfies the requirement dbus-python (from displaycal) (from versions: none)
       > ERROR: No matching distribution found for dbus-python
```

when building the derivation file (here-called displaycal.nix) like this `nix-build -E "with import <nixpkgs> {}; callPackage ./displaycal.nix { inherit (pkgs.python310Packages) PyChromecast build buildPythonApplication certifi dbus-python distro fetchPypi send2trash wxPython_4_1; }"`

I am most likely doing something wrong

**Comment by toastal** [comment]
@TheSirC did the patch I leave happen to help??

**Comment by toastal** [comment]
@Kranzes I found out how to overcome the issues with the file explorer with a new patch:

```diff
-diff --git a/pkgs/applications/graphics/displaycal/default.nix b/pkgs/applications/graphics/displaycal/default.nix
-index 646751b0577..89f38ce03e5 100644
---- a/pkgs/applications/graphics/displaycal/default.nix
-+++ b/pkgs/applications/graphics/displaycal/default.nix
-@@ -11,6 +11,8 @@
- , distro
- , makeWrapper
- , argyllcms
-+, gtk3
-+, gsettings-desktop-schemas
- }:
- 
- buildPythonApplication rec {
-@@ -53,8 +55,9 @@ buildPythonApplication rec {
- 
-     for binary in "$out/bin/"*; do
-       wrapProgram "$binary" \
-+        --prefix PATH : "${lib.makeBinPath [ argyllcms ]}" \
-         --prefix PYTHONPATH : "$PYTHONPATH" \
--        --prefix PATH : ${lib.makeBinPath [ argyllcms ]}
-+        --prefix XDG_DATA_DIRS : "${gsettings-desktop-schemas}/share/gsettings-schemas/${gsettings-desktop-schemas.name}:${gtk3}/share/gsettings-schemas/${gtk3.name}"
-     done
-   '';
- 
```

**Comment by sifmelcara** [comment]
Thanks for working on this! However it does not seem to work? (Not sure if it is my environment's issue) I get the following error when click `Calibrate & profile`

```
b'Traceback (most recent call last):\n File "<string>", line 1, in <module> \ n File 
"/nix/store/2qm93yp3zh5vy9dibghzjlxn154sh53f-displaycal-3.9.7/1ib/python3.10/site-packages/DisplayCAL/wxMeasureFrame.py", line 9, in 
<module> \n from DisplayCAL import config\nModuleNotFoundError: No module named \'DlisplayCAL\''
```
(I ran it with `nix run github:NixOS/nixpkgs/68c69155099ad999a5e08a50df7272a39d9b405c#displaycal`)

display in 21.11 channel works fine for me

**Comment by toastal** [comment]
@sifmelcara Did you try the two patches I tried enclosing but are left as comments?

**Comment by sifmelcara** [comment]
@toastal Thanks they works, didn't realize they also fixes python path

**Comment by toastal** [comment]
Since then, 3.9.8 https://github.com/eoyilmaz/displaycal-py3/releases/tag/3.9.8

The new hash: `sha256-fT+WFzbGegdvIX4khpAmtO4qpgjxorZw1ai1NCdYL/4=`

**Comment by toastal** [comment]
```
{ lib
, buildPythonApplication
, fetchPypi
, xorg
, certifi
, dbus-python
, build
, wxPython_4_1
, send2trash
, PyChromecast
, distro
, makeWrapper
, argyllcms
, gtk3
, gsettings-desktop-schemas
}:

buildPythonApplication rec {
  pname = "displaycal";
  version = "3.9.10";

  src = fetchPypi {
    pname = "DisplayCAL";
    inherit version;
    sha256 = "sha256-oDHDVb0zuAC49yPfmNe7xuFKaA1BRZGr75XwsLqugHs=";
  };

  propagatedBuildInputs = [
    certifi
    dbus-python
    build
    wxPython_4_1
    send2trash
    PyChromecast
    distro
  ];

  buildInputs = with xorg; [
    libX11
    libXxf86vm
    libXext
    libXinerama
    libXrandr
  ];

  pythonImportsCheck = [ "DisplayCAL" ];

  doCheck = false; # Tests try to access an X11 session.

  # no idea why it looks there - symlink .json lang (everything)
  postInstall = ''
    for x in $out/share/DisplayCAL/*; do
      ln -s $x $out/lib/python3.10/site-packages/DisplayCAL
    done

    for binary in "$out/bin/"*; do
      wrapProgram "$binary" \
        --prefix PATH : "${lib.makeBinPath [ argyllcms ]}" \
        --prefix PYTHONPATH : "$PYTHONPATH" \
        --prefix XDG_DATA_DIRS : "${gsettings-desktop-schemas}/share/gsettings-schemas/${gsettings-desktop-schemas.name}:${gtk3}/share/gsettings-schemas/${gtk3.name}"
    done
  '';

  meta = with lib; {
    description = "DisplayCAL Modernization Project (Migrated to Python 3)";
    homepage = "https://github.com/eoyilmaz/displaycal-py3";
    license = licenses.gpl3Plus;
    maintainers = with maintainers; [ kranzes toastal ];
  };
}
```

@Kranzes here's the version for 3.9.10. I included myself as a maintainer because I need to use and test this a lot more than yourself. Can we please get an update and try to get this merged??

**Comment by Kranzes** [comment]
Result of `nixpkgs-review pr 187303` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>displaycal</li>
  </ul>
</details>

**Review by toastal — APPROVED** [review]

File: pkgs/applications/graphics/displaycal/default.nix
Comment (amjoseph-nixpkgs) [inline pkgs/applications/graphics/displaycal/default.nix:69]: Recursive `description` is recursive?

```suggestion
    description = "Display calibration tool *migrated to Python 3)";
```

Comment (toastal) [reply pkgs/applications/graphics/displaycal/default.nix:69]: That's a reasonable point. DisplayCAL is a long-established name in the space it exists, so the migration of it just referencing it by name make sense in the context, but I can certainly see how a more generic name is better suited for users unfamiliar with the project--especially since if you search by name DisplayCAL should already show up.

Comment (toastal) [reply pkgs/applications/graphics/displaycal/default.nix:69]: Aside: `ikhal` via `khal` can get you a TUI calendar compatible with `vdirsyncer` (with the caveat that it's Python so it's slow and breaks frequently to where multiple people have overlays to disable `doCheck` because CI passing isn't a requirement for their upstream it seems).

Comment (toastal) [reply pkgs/applications/graphics/displaycal/default.nix:69]: The old Python 2.x version used this description.

> Display Calibration and Characterization powered by Argyll CMS

Though I would suggest using sentence case over random capital letters: "Display calibration and characterization powered by Argyll CMS"

File: pkgs/top-level/all-packages.nix
Comment (FRidh) [inline pkgs/top-level/all-packages.nix:25256]: Please remove here `python3Packages.`

File: pkgs/applications/graphics/displaycal/default.nix
Comment (FRidh) [inline pkgs/applications/graphics/displaycal/default.nix:1]: Take `python3` as parameter instead of all the Python packages/functions.

File: pkgs/applications/graphics/displaycal/default.nix
Comment (FRidh) [inline pkgs/applications/graphics/displaycal/default.nix:28]: can be removed

File: pkgs/applications/graphics/displaycal/default.nix
Comment (FRidh) [inline pkgs/applications/graphics/displaycal/default.nix:60]: You can set `makeWrapperArgs` for a Python package. It takes a list of arguments you would give to `wrapProgram` (such as `--prefix ..`)

**Review by FRidh — CHANGES_REQUESTED** [review]

File: pkgs/applications/graphics/displaycal/default.nix
Comment (toastal) [reply pkgs/applications/graphics/displaycal/default.nix:1]: Might we want the Python 3.10 specifically since the derivation is doing some linking to the versioned directory?

File: pkgs/applications/graphics/displaycal/default.nix
Comment (toastal) [reply pkgs/applications/graphics/displaycal/default.nix:60]: Is that an attribute in the set for `buildPythonApplication`

Comment (FRidh) [reply pkgs/applications/graphics/displaycal/default.nix:60]: It's an environment variable that the `wrapPython` hook listens to.

File: pkgs/applications/graphics/displaycal/default.nix
Comment (FRidh) [inline pkgs/applications/graphics/displaycal/default.nix:57]:
```suggestion
      ln -s $x $out/${python3.sitePackages}/DisplayCAL
```

File: pkgs/applications/graphics/displaycal/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/graphics/displaycal/default.nix:58]: Can we use the gappsWrapper instead?

Comment (Kranzes) [reply pkgs/applications/graphics/displaycal/default.nix:53]: Is that the equivalent?

Comment (Kranzes) [reply pkgs/applications/graphics/displaycal/default.nix:53]: Doesn't seem to be the case. It complains about missing schemas and crashes when trying to open file browser.

**Comment by davidak** [comment]
The commit message should be "displaycal: init at 3.9.10" to follow CONTRIBUTING.

I will test later. Thanks for the efforts!

File: pkgs/applications/graphics/displaycal/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/graphics/displaycal/default.nix:12]:
```suggestion
  version = "3.9.10";
  format = "setuptools";
```
at some point this will become mandatory :)

**Comment by toastal** [comment]
Latest builds have worked for me

**Comment by toastal** [comment]
Current pushes do to missing libraries from the `PYTHONPATH`

**Review by 06kellyjac — APPROVED** [review]

File: pkgs/applications/graphics/displaycal/default.nix
Comment (06kellyjac) [inline pkgs/applications/graphics/displaycal/default.nix:66]: why did you remove yourself as a maintainer @Kranzes ? :sweat_smile:

File: pkgs/applications/graphics/displaycal/default.nix
Comment (mweinelt) [reply pkgs/applications/graphics/displaycal/default.nix:63]: We prefer the repo URl anyway.

File: pkgs/applications/graphics/displaycal/default.nix
Comment (Kranzes) [reply pkgs/applications/graphics/displaycal/default.nix:66]: I can't properly test the program without the hardware, @toastal has the hardware and he personally requested me to package it.

**Review by toastal — APPROVED** [review]
Current state builds an works on my machine with colormunki colorimeter.

**Comment by Artturin** [comment]
it starts

File: pkgs/applications/graphics/displaycal/default.nix
Comment (Atemu) [inline pkgs/applications/graphics/displaycal/default.nix:26]:
```
       error: undefined variable 'PyChromecast'

       at /Users/atemu/.cache/nixpkgs-review/pr-218309-13/nixpkgs/pkgs/applications/graphics/displaycal/default.nix:32:5:

           31|     distro
           32|     PyChromecast
             |     ^
           33|     send2trash
```



```Suggestion
  propagatedBuildInputs = with python3Pkgs; [
```

Comment (Atemu) [reply pkgs/applications/graphics/displaycal/default.nix:26]: Please fixup, this currently breaks eval/nixpkgs-review.

Comment (toastal) [reply pkgs/applications/graphics/displaycal/default.nix:26]: There is no `python3Pkgs`. There is `python310Pkgs`. I just ran a `nixpkgs-review` successfully @ https://github.com/NixOS/nixpkgs/pull/219584#pullrequestreview-1325241857 and I'm on a commit with these changes. Can you give more info because this change doesn't even compile because of `python3Pkgs`?

Comment (Kranzes) [reply pkgs/applications/graphics/displaycal/default.nix:26]: I just build it just fine directly from master.

Comment (Artturin) [reply pkgs/applications/graphics/displaycal/default.nix:26]: evals ok `nix-repl> displaycal
«derivation /nix/store/60kf5hp91k1kh5wbhc246hnldlk32rz7-displaycal-3.9.10.drv»`

python3.pkgs is the same thing as python3Packages

Comment (Atemu) [reply pkgs/applications/graphics/displaycal/default.nix:26]: Odd, I cannot reproduce the issue myself with manual evals either. Could you try `nixpkgs-review pr 218309`? It should fail after the git merge command.

Comment (Atemu) [reply pkgs/applications/graphics/displaycal/default.nix:26]: Alright, narrowed it down to something else. Really looked like this was the cause though oddly. Yay intransparent error messages?

**Pull request merged**
